test/standalone/test_rpc_proxy.py

```diff
@@ -222,85 +223,119 @@ def test_multiple_calls_to_result(container_factory, rabbit_config):
+        self, container_factory, rabbit_manager, rabbit_config, rpc_proxy
+    ):
+        block = Event()
 
-        @rpc
-        def method(self, arg):
-            self.hook()
-            block.wait()
-            return arg
+        class Service(object):
+            name = "service"
 
-    container = container_factory(ExampleService, rabbit_config)
-    container.start()
+            @rpc
+            def method(self, arg):
+                block.wait()
+                return arg
 
-    vhost = rabbit_config['vhost']
+        container = container_factory(Service, rabbit_config)
+        container.start()
 
-    # get exampleservice's queue consumer connection while we know it's the
-    # only active connection
-    connections = get_rabbit_connections(vhost, rabbit_manager)
-    assert len(connections) == 1
-    container_connection = connections[0]
-
-    with ServiceRpcProxy('exampleservice', rabbit_config) as proxy:
-
-        # grab the proxy's connection too, the only other connection
-        connections = get_rabbit_connections(vhost, rabbit_manager)
-        assert len(connections) == 2
-        proxy_connection = [
-            conn for conn in connections if conn != container_connection
-        ][0]
-
-        counter = itertools.count(start=1)
-
-        class ConnectionStillOpen(Exception):
-            pass
-
-        @retry(for_exceptions=ConnectionStillOpen, delay=0.2)
-        def wait_for_connection_close(name):
-            connections = get_rabbit_connections(vhost, rabbit_manager)
-            for conn in connections:
-                if conn['name'] == name:
-                    raise ConnectionStillOpen(name)  # pragma: no cover
-
-        def cb(args, kwargs, res, exc_info):
-            # trigger a disconnection on the second call.
-            # release running workers once the connection has been closed
-            count = next(counter)
-            if count == 2:
-                rabbit_manager.delete_connection(proxy_connection['name'])
-                wait_for_connection_close(proxy_connection['name'])
-                block.send(True)
-                return True
-
-        # attach a callback to `hook` so we can close the connection
-        # while there are requests in-flight
-        with wait_for_call(ExampleService, 'hook', callback=cb):
-
-            # make an async call that runs for some time
-            async_call = proxy.method.call_async("hello")
-
-            # make another call that will trigger the disconnection;
-            # expect the blocking proxy to raise when the service reconnects
-            with pytest.raises(RpcConnectionError):
-                proxy.method("hello")
+        # make an async call that will block,
+        # wait for the worker to have spawned
+        with wait_for_call(container, 'spawn_worker'):
+            res = rpc_proxy.service.method.call_async('msg1')
+
+        # fake a socket error that causes the proxy to reconnect
```

Comment:
the original test actually disconnected. not sure about the fidelity of just mocking the expected side-effect

Reply:
The original test isn't particularly representative of real behaviour -- closing the connection from the broker API actually sends a consumer-cancel rather than yanking the connection.

We _could_ put toxiproxy in here, but it seems like overkill if we trust that kombu will throw ConnectionErrors when something goes awry with the connection.

Reply:
How hard/annoying is it to add a toxiproxi test? Not so keen on relying on mocks here, esp as we’re looking at upgrading or even removing kombu

Reply:
That's a good point. It's probably not too hard. Will try to add it

Reply:
Good call on adding this -- the behaviour on a real disconnection is different, because kombu keeps the state of the connection and we use it to determine whether the proxy has been previously stopped.

I've updated the test, although the whole thing will be made moot by the fix for https://github.com/nameko/nameko/issues/359